### ISSUE TITLE
NAACL-2019-BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding

### ISSUE THREAD

**Issue by BrambleXu** [BrambleXu/knowledge-graph-learning#103]
**一句话总结：**

双向Transformer, 通过Masked LM，做到context aware。效果拔群。

(NAACL-2019  Best Long Paper)

**资源：**

- [pdf](<https://arxiv.org/abs/1810.04805>)
- code: [tensorflow](<https://github.com/google-research/bert>), [pytorch](<https://github.com/huggingface/pytorch-pretrained-BERT>)
- [paper-with-code](<https://paperswithcode.com/paper/bert-pre-training-of-deep-bidirectional>)

**关键字：**

- dataset: 
- BERT, which stands for Bidirectional Encoder Representations from
  Transformers
- BERT is designed to pre-train deep bidirectional representations by jointly conditioning on both left and right context in all layers.
- As a result, the pre-trained BERT representations can be fine-tuned with just one additional output layer to create state-of-the art models for a wide range of tasks, such as question answering and language inference, without substantial task-specific architecture
  modifications.

**笔记：**

#  1 Introduction

There are two existing strategies for applying pre-trained language representations to downstream tasks: feature-based and fine-tuning

- The feature-based approach, such as ELMo (Peters et al., 2018), uses tasks-specific architectures that include the pre-trained representations as additional features
- The fine-tuning approach, such as the Generative Pre-trained Transformer (OpenAI GPT) (Radford et al., 2018), introduces minimal
  task-specific parameters, and is trained on the downstream tasks by simply fine-tuning the pretrained parameters.

Both approaches share the same objective function during pre-training, where they use unidirectional language models to learn general language representations (这两种方法有相同的目标函数)

本文认为这种做法限制了预训练模型的潜在能力，尤其是fine-tuning approach。

最主要的缺陷是这些模型是unidirectional的，这个限制了模型在预训练时的选择。比如在OpenAI GPT中，作者用了left-to-right架构，每个token只能添加到前一个token (in the self-attention layers of the Transfomer)。这种限制对于sentence level task是sub-optimal的，对于token-level的效果也会大打折扣（and could be devastating when applying fine-tuning based approaches to token-level tasks such as SQuAD question answering）。所以考虑双向是很关键的。

这篇文章提出了BERT，用来解决上面提到的unidirectional constraints，通过提出一个新的pre-training objective: the "masked language model" (MLM)。这个MLM随机把input里的token进行mask，而我们的目标则是预测original vocabulary id of the masked word based only on its context. 基于这样的设置，MLM能让模型关注于token左右两边的context。

除了MLM，还提出了一个“next sentence prediction” task, that jointly pre-trains text-pair representations. 

2 Related Work

2.1 Feature-based Approaches

2.2 Fine-tuning Approaches

2.3 Transfer Learning from Supervised Data

# 3 BERT

## 3.1 Model Architecture

BERT’s model architecture is a multi-layer bidirectional Transformer encoder based on the original implementation #102 . 这里用的T和102论文里用的完全一样。

we denote the number of layers (i.e., Transformer blocks) as L, the hidden size as H, and the number of self-attention heads as A. 层数是L，隐藏神经元数量是H，self-attention head的数量是A。In all cases we set the feed-forward/filter size to be 4H, i.e., 3072 for the H = 768 and 4096 for the H = 1024. We primarily report results on two model sizes:

- BERT_BASE: L=12, H=768, A=12, Total Parameters=110M
- BERTLARGE: L=24, H=1024, A=16, Total Parameters=340M

为了和OpenAi GPT进行比较，BERT_BASE选择了一样大小的size。

![image-20190411134447044](https://ws1.sinaimg.cn/large/006tNc79gy1g1ylgqmk0vj30t70aygoy.jpg)

## 3.2 Input Representation

解释一下这篇文章里的专业名字含义

- “sentence”: an arbitrary span of contiguous text, rather than an actual linguistic sentence.
- “sequence”: refers to the input token sequence to BERT, which may be a single sentence or two sentences packed together.
- token：representation is constructed by summing the corresponding token, segment and position embeddings. 

具体使用到的输入：

![image-20190411134511729](https://ws2.sinaimg.cn/large/006tNc79gy1g1ylh6elm6j30tk0apwgc.jpg)

(这个segament embedding是什么？下面解释了，是sentence embedding)

- WordPiece embeddings (Wu et al., 2016) with a 30,000 token vocabulary。 We denote split word pieces with ##
- We use learned positional embeddings with supported sequence lengths up to 512 tokens
- The first token of every sequence is always the special classification embedding ([CLS]). The final hidden state (i.e., output of Transformer) corresponding to this token is used as the aggregate sequence representation for classification tasks. For nonclassification tasks, this vector is ignored. 每个sequence开始的token都是`[CLS]`
- Sentence pairs are packed together into a single sequence. We differentiate the sentences in two ways. First, we separate them with
  a special token ([SEP]). Second, we add **a learned sentence A embedding** to every token of the first sentence and a sentence B embedding to every token of the second sentence
- For single-sentence inputs we only use the sentence A embeddings.

(*a learned sentence A embedding*, 那么这里的sentence emebdding是从哪里learn来的？)

## 3.3 Pre-training Tasks

We do not use traditional left-to-right or right-to-left language models to pre-train BERT. Instead, we pre-train BERT using two novel unsupervised prediction tasks, described in this section. 通过两个新的unsupervised prediction 任务来进行预训练。

### 3.3.1 Task #1: Masked LM

直觉上我们认为deep bidirectional model比left-to-right model or the shallow concatenation of a left-to-right and right-toleft model 效果更好。Unfortunately, standard conditional
language models can only be trained left-to-right or right-to-left, since bidirectional conditioning would allow each word to indirectly “see itself” in a multi-layered context. 但是双向的话，因为是multi-layerd context, 每个token都能间接地看到"自己"。

为了训练deep bidirectional representation, we take a straightforward approach of masking some percentage of the input tokens at random, and then predicting only those masked tokens。(等等，这么看的话，tacred-relation里的设置`words = [word_dropout(sent, self.opt['word_dropout']) for sent in batch[0]]`可能也是出于这样的原因。嘛，能增强健壮性到时比较直觉)。Refer to this procedure as a “masked LM” (MLM).

In this case, the final hidden vectors corresponding to the mask tokens are fed into an output softmax over the vocabulary, as in a standard LM. In all of our experiments, we mask 15% of all WordPiece tokens in each sequence at random. (之所以设定15%这么高，可能也是因为使用了WordPiece tokens吧。sentence通常会比较长)。In contrast to denoising auto-encoders (Vincent et al., 2008), we only predict the masked words rather than reconstructing the entire input. 只预测masked的单词。如果是预测整个input的话，和auto-encoder就差不多了。

这种方式能让我们获得一个bidirectional pre-trained model,但是有两个缺点。

第一个缺点，we are creating a mismatch between pre-training and finetuning, since the [MASK] token is never seen during fine-tuning. 

为了解决这个问题，To mitigate this, we do not always replace “masked” words with the actual [MASK] token. Instead, the training data generator chooses 15% of tokens at random, e.g., in the sentence my dog is hairy it chooses hairy. It then performs the following procedure:

- 并不总是把chosen work用[MASK]来表示。Rather than always replacing the chosen
  words with [MASK], the data generator will do the following:
  - 80% of the time: Replace the word with the [MASK] token, e.g., my dog is hairy →
    my dog is [MASK]
  - 10% of the time: Replace the word with a random word, e.g., my dog is hairy → my
    dog is apple
  - 10% of the time: Keep the word unchanged, e.g., my dog is hairy → my dog is hairy. The purpose of this is to bias the representation towards the actual observed word.

(等等，那这里的时间具体指的是什么？epoch？)

Because random replacement only occurs for 1.5% of all tokens (i.e., 10% of 15%), this does not seem to harm the model’s language understanding capability. 

(为什么是1.5%? 哦哦，首先是只选择15%的的token进行random mask, 然后对于random masked token, 其中只有10%保持原来的真实token。所以总共只有1.5%)

第二个缺点，using an MLM is that only 15% of tokens are predicted in each batch, which suggests that more pre-training steps may be required for the model to converge. In Section 5.3 we demonstrate that MLM does converge marginally slower than a left-to-right model
(which predicts every token), but the empirical improvements of the MLM model far outweigh the increased training cost. 看来这个训练时间长的缺点没有办法很好解决啊。作者说虽然时间长，但是效果好。

### 3.3.2 Task #2: Next Sentence Prediction

Many important downstream tasks such as Question Answering (QA) and Natural Language Inference (NLI) are based on understanding the relationship between two text sentences. ええ～、そうなんだ～。QA和NLI主要是想理解两个text sentences之间的关系，而这种关系，无法直接被language modeling博捉到。为了训练一个能理解 sentence relation的模型，we pre-train a binarized next sentence prediction task that can be trivially generated from any monolingual corpus. 

这个任务其实也比较简单，和上面的MLM有点像。每次抽选A和B两个句子，B有50%的几率，是A的下一句。B的另外50%的机会，是从corpus中随机选取的句子。 when choosing the sentences A and B for each pretraining example, 50% of the time B is the actual next sentence that follows A, and 50% of the time it is a random sentence from the corpus. For example:

```
Input = [CLS] the man went to [MASK] store [SEP] he bought a gallon [MASK] milk [SEP]

Label = IsNext


Input = [CLS] the man [MASK] to the store [SEP] penguin [MASK] are flight ##less birds [SEP]

Label = NotNext
```

We choose the NotNext sentences completely at random, and the final pre-trained model achieves 97%-98% accuracy at this task. Despite its simplicity, we demonstrate in Section 5.1 that pretraining towards this task is very beneficial to both QA and NLI. 注意，这种猜句子的方式，对于QA，NLI这样需要理解句子的任务来说很有用。

## 3.4 Pre-training Procedure

Pre-training corpus we use the concatenation of BooksCorpus (800M words) (Zhu et al.,
2015) and English Wikipedia (2,500M words). For Wikipedia we extract only the text passages
and ignore lists, tables, and headers.

It is critical to use a document-level corpus rather than a shuffled sentence-level corpus such as the Billion Word Benchmark (Chelba et al., 2013) in order to extract long contiguous sequences.选择document-level corpus很重要。

给`Next Sentence Prediction`准备数据集。we refer to as “sentences” even though they are
typically much longer than single sentences (but can be shorter also). The first sentence receives the A embedding and the second receives the B embedding. 50% of the time B is the actual next sentence that follows A and 50% of the time it is a random sentence, which is done for the “next sentence prediction” task. They are sampled such that the combined length is ≤ 512 tokens. The LM masking is applied after WordPiece tokenization with a uniform masking rate of 15%, and no special consideration given to partial word pieces.

这一段是介绍各种参数的，需要复现的时候再仔细看吧。Train with batch size of 256 sequences (256 sequences * 512 tokens = 128,000 tokens/batch) for 1,000,000 steps, which is approximately 40 epochs over the 3.3 billion word corpus. We use Adam with learning rate of 1e-4, β1 = 0.9, β2 = 0.999, L2 weight decay of 0.01, learning rate warmup over the first 10,000 steps, and linear decay of the learning rate. We use a dropout probability of 0.1 on all layers. We use a gelu activation (Hendrycks and Gimpel, 2016) rather than the standard relu, following OpenAI GPT. The training loss is the sum of the mean masked LM likelihood and mean next sentence prediction likelihood. 损失函数是sum of the mean masked LM likelihood and mean next sentence prediction likelihood.

## 3.5 Fine-tuning Procedure

For sequence-level classification tasks, BERT fine-tuning is straightforward. In order to obtain
a fixed-dimensional pooled representation of the input sequence, we take the final hidden state (i.e., the output of the Transformer) for the first token in the input, which by construction corresponds to the the special [CLS] word embedding. 

For spanlevel and token-level prediction tasks, the above procedure must be modified slightly in a taskspecific manner. 需要注意的就是这一句，根据具体任务的不同，维度要进行适度的变更。

黄金配置

![image-20190411153803836](https://ws2.sinaimg.cn/large/006tNc79gy1g1yoqlnaj8j30ee06o75b.jpg)

We also observed that large data sets (e.g., 100k+ labeled training examples) were far less
sensitive to hyperparameter choice than small data sets. Fine-tuning is typically very fast, so it is reasonable to simply run an exhaustive search over the above parameters and choose the model that performs best on the development set

## 3.6 Comparison of BERT and OpenAI GPT

- GPT uses a sentence separator ([SEP]) and classifier token ([CLS]) which are only introduced at fine-tuning time; BERT learns [SEP], [CLS] and sentence A/B embeddings during pre-training. 

# 4 Experiments

![image-20190411154323270](https://ws3.sinaimg.cn/large/006tNc79gy1g1yow59pq2j30pu0ontem.jpg)

![image-20190411154407182](https://ws2.sinaimg.cn/large/006tNc79gy1g1yowwtsxdj30pg0b3n0m.jpg)

![image-20190411154441452](https://ws3.sinaimg.cn/large/006tNc79gy1g1yoxi96pjj30ct094abc.jpg)

# 5 Ablation Studies

An ablation study typically refers to removing some “feature” of the model or algorithm, and seeing how that affects performance. [Ablation Studies](<https://www.quora.com/In-the-context-of-deep-learning-what-is-an-ablation-study>)

## 5.1 Effect of Pre-training Tasks

这里认为BERT之所以有这么好的效果，主要是因为使用了masked LM pre-training。为了验证这一点，进行了以下的对比实验。

1. No NSP: A model which is trained using the “masked LM” (MLM) but without the “next
   sentence prediction” (NSP) task. 只用一个MLM任务训练，不使用NSP任务。
2. LTR & No NSP: A model which is trained using a Left-to-Right (LTR) LM, rather than an MLM. In this case, we predict every input word and do not apply any masking. 使用LTR LM模型进行训练，并且不使用masking。这个是为了和OpenAI GPT做对比，区别只是用的数据集不一样。

![image-20190411163324386](https://ws1.sinaimg.cn/large/006tNc79gy1g1yqc6xoxaj30fg0crgo1.jpg)

- 去除了NSP后，效果下降还是挺明显的。
- 然后比较 “No NSP”to “LTR & No NSP”. 前者使用了MLM，后者没有用mask。对于QNLI, SQUAD这两个数据集来说，发现效果下降的非常明显。The LTR model performs
  worse than the MLM model on all tasks, with extremely large drops on MRPC and SQuAD.

5.2 Effect of Model Size

5.3 Effect of Number of Training Steps

5.4 Feature-based Approach with BERT
